fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/ratefactor/RateFactorFunctions.java

```diff
@@ -0,0 +1,134 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanproduct.ratefactor;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.fineract.infrastructure.core.service.DateUtils;
+import org.apache.fineract.organisation.monetary.domain.MoneyHelper;
+import org.apache.fineract.portfolio.common.domain.DaysInYearType;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepaymentScheduleInstallment;
+
+public class RateFactorFunctions {
+
+    public static final Integer PRECISION = 8;
+
+    protected RateFactorFunctions() {}
+
+    /**
+     * To calculate the monthly payment, we first need to calculate something called the Rate Factor. We're going to be
+     * using simple interest. The Rate Factor for simple interest is calculated by the following formula:
+     *
+     *
+     * R = 1 + (r * d / y)
+     *
+     * @param interestRate
+     *            (r)
+     * @param daysInPeriod
+     *            (d)
+     * @param daysInYear
+     *            (y)
+     */
+    public static BigDecimal rateFactor(final BigDecimal interestRate, final Long daysInPeriod, final Integer daysInYear,
+            final Integer precision) {
+        final BigDecimal daysPeriod = BigDecimal.valueOf(daysInPeriod);
+        BigDecimal daysYear = BigDecimal.valueOf(daysInYear);
+        final MathContext mc = new MathContext(precision, MoneyHelper.getRoundingMode());
+
+        return BigDecimal.ONE.add(interestRate.multiply(daysPeriod.divide(daysYear, mc), mc), mc);
+    }
+
+    /**
+     * To calculate the function value for each period, we are going to use the next formula:
+     *
+     * fn = 1 + fnValueFrom * rateFactorEnd
+     *
+     * @param periodFrom
+     *
+     * @param periodEnd
+     *
+     * @param mathContext
+     *
+     */
+    public static RateFactorPeriodData fnValue(RateFactorPeriodData periodFrom, final RateFactorPeriodData periodEnd,
+            final MathContext mc) {
+        BigDecimal fnValue = BigDecimal.ONE;
+        if (periodEnd.getPeriod() > 1) {
+            fnValue = BigDecimal.ONE.add(periodFrom.getFnValue().multiply(periodEnd.getRateFactor(), mc));
+        }
+        periodEnd.setFnValue(fnValue);
+        return periodEnd;
+    }
+
+    /**
+     *
+     *
+     * emi = (R^n * principal) / fnValue
+     *
+     * @param periods
+     *
+     * @param principal
+     *
+     * @param mathContext
+     *
+     */
+    public static BigDecimal calculateEMI(final List<RateFactorPeriodData> periods, final BigDecimal principal, final MathContext mc,
+            final Integer precision) {
+        final BigDecimal lastFnValue = periods.get(periods.size() - 1).getFnValue();
+        BigDecimal rnValue = BigDecimal.ONE;
+        for (RateFactorPeriodData rateFactorPeriod : periods) {
+            if (rateFactorPeriod.getPeriod() == 1) {
```

Comment:
This condition is unnecessary:
`rnValue = rnValue.multiply(rateFactorPeriod.getRateFactor(), mc);` gives you the right value even it is the 1st period.